suse_migration_services/units/product_setup.py

```diff
@@ -66,53 +68,66 @@ def main():
+                log.error(message)
+                raise DistMigrationProductSetupException(message)
         log.info('Updating Base Product to be suitable for migration')
-        update_baseproduct(baseproduct)
+        update_baseproduct(baseproduct.prod_filename)
     except Exception as issue:
         message = 'Base Product update failed with {0}'.format(issue)
         log.error(message)
         raise DistMigrationProductSetupException(message)
 
 
 def get_baseproduct(products_metadata):
-    baseproduct_filename = []
-    xml = ElementTree()
-    for product_name in os.listdir(products_metadata):
-        product = os.sep.join(
-            [products_metadata, product_name]
-        )
-        xml.parse(product)
-        register_sections = xml.findall('register')
-        for register in register_sections:
-            flavor = register.findall('flavor')
-            if not flavor:
-                baseproduct_filename.append(product)
+    base_product_type = namedtuple(
+        'base_product_type', ['count', 'prod_filename']
+    )
+    prod_filenames = glob.glob(
+        os.path.join(products_metadata, '*.prod')
+    )
+    count = 0
+    name = None
+    try:
+        xml = ElementTree()
+        for prod_filename in prod_filenames:
+            xml.parse(prod_filename)
+            register_sections = xml.findall('register')
+            for register in register_sections:
+                flavor = register.findall('flavor')
+                if not flavor:
+                    count = count + 1
+                    name = prod_filename
 
-    return baseproduct_filename
+        return base_product_type(count, name)
+    except Exception as issue:
+        message = ('Getting base product for product '
+                   'update failed with {0}'.format(issue))
+        log.error(message)
+        raise DistMigrationProductSetupException(message)
 
 
-def update_baseproduct(baseproduct):
-    xml = ElementTree()
-    xml.parse(baseproduct)
-    register_sections = xml.findall('register')
-    for register in register_sections:
-        target_sections = register.findall('target')
-        for target in target_sections:
-            register.remove(target)
-    xml.write(
-        baseproduct, encoding="UTF-8", xml_declaration=True
-    )
+def update_baseproduct(baseproduct_prod_filename):
+    try:
+        xml = ElementTree()
+        xml.parse(baseproduct_prod_filename)
+        register_sections = xml.findall('register')
+        for register in register_sections:
+            target_sections = register.findall('target')
+            for target in target_sections:
+                register.remove(target)
+        xml.write(
+            baseproduct_prod_filename, encoding="UTF-8", xml_declaration=True
+        )
+    except Exception as issue:
+        message = 'Base product update failed with {0}'.format(issue)
+        log.error(message)
+        raise DistMigrationProductSetupException(message)
```

Comment:
See comments about "best effort"

Reply:
Regarding the objection on raise conditions I disagree with you here. I think they are properly set.
If any code in the try block failed this means there was an error on product data backup or on reading / handling product files. We should not continue the migration if there was an error. The best effort approach does not apply here because we just move the error to the subsequent zypper call.

The other raise condition is if the base product counter is != 1. In this case I also think we should not continue the migration. The result of zypper is basically undefined if there are more than one base products and we should raise before calling zypper.

All of that of course only if the way we detect the base product is correct. So far the lookup for the only one with no flavour set seems correct.

Thus I would not change the raise conditions. At this point we still can safely stop and haven't destroyed the system. The level of strictness is imho ok and preferable

Reply:
sorry I think I misread it. I just saw there is a raise call in get_baseproduct(). I agree that method should not raise, just log and always return with a result. The result is evaluated and that could lead to raise

Reply:
Same for delete_target_registration. That method doesn't need to try / catch. It's result is covered by the overall try block where we call it.

Reply:
I think that was the part Robert was referring too and that would make sense to me too

Reply:
As delete_target_registration raises with the same exception than the overall try block it got called it doesn't make sense. So I suggest to not raise in delete_target_registration

less code is better